### PR TITLE
lxd: use LXD-compatible instance names

### PR DESCRIPTION
**Pull request by mr-cal**
- [X] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----
### Overview
The provided name for a LXD executor object is converted to comply with [LXD naming conventions](https://linuxcontainers.org/lxd/docs/master/instances/) for instances.

### Problem
LXC has a 63 character limit on instance names.  [This snapcraft commit](https://github.com/snapcore/snapcraft/commit/21cb1ab9ae26cf3322356c190eb555ed05f2d498) adds 18-20 characters to the image name.  This is blocking snaps with long names from upgrading to core22.

Charmcraft has the [same issue](https://github.com/canonical/charmcraft/issues/629).

#### Snapcraft naming convention
```
snapcraft-<project-name>-on-<build-on>-for-<build-for>-<project-path-inode>
└  10   ┘└   2 - 41     ┘└   8-11    ┘└    9-12       ┘└       10         ┘
```
Min size: 10 + 2 + 8 + 9 + 10 = 39
Max size: 10 + 41 + 11 + 12 + 10 = 84

#### Charmcraft naming convention
```
charmcraft-<project-name>-<project-path-inode>-<bases-index>-<build-on-index>-<build-for>
└  11   ┘└    2 - ∞      ┘└       10          ┘└     2      ┘└      2        ┘└    7    ┘
```
Min size:   11 + 2 + 10 + 2 + 2 + 7 = 34
Max size: unbound

### Solution
`craft-providers` is responsible for complying with each provider’s specific requirements.
For LXD, `craft-providers` will:
1. truncate the first 40 characters name
2. remove forbidden characters
3. append a hash of the name
```
<truncated-name>-<hash-of-name>
└    1 - 40    ┘1└     20     ┘
```
Min size: 1 + 1 + 20 = 22
Max size: 40 + 1 + 20 = 61

Note: This is an internal change.  Applications using `craft-providers` should continue to choose their own unique names for instances.  The only restriction is that the name must contain at least one alphanumeric character.

LP: #[1981205](https://bugs.launchpad.net/snapcraft/+bug/1981205)
(CRAFT-1235)

### CODE REVIEW

File: craft_providers/lxd/lxd_instance.py
Comment (cmatsuoka) [inline craft_providers/lxd/lxd_instance.py:124]: Shouldn't we use the new instance name only if the old-style name is invalid or too large, to keep using existing instances?

Comment (mr-cal) [reply craft_providers/lxd/lxd_instance.py]: Good catch, I added that in.

File: craft_providers/lxd/lxd_instance.py
Comment (cmatsuoka) [inline craft_providers/lxd/lxd_instance.py:120]: This test doesn't seem correct, truncated names are clipped at 40 char size but previous valid names could have up to 64?

File: craft_providers/lxd/lxd_instance.py
Comment (facundobatista) [inline craft_providers/lxd/lxd_instance.py:121]: I think we should do this validation before truncating.

If the proposed name by the app is valid and, say, 55c long, it should be ok and use it.

So, removing non valid chars and trimming digits and dashes it's a mandatory step (so we avoid invalid names), but then if the len of the resulting thing is < 62, just use it! if it's > 62, truncate to 40ish and append hash, yes.

What do you think?

File: craft_providers/lxd/lxd_instance.py
Comment (facundobatista) [inline craft_providers/lxd/lxd_instance.py:126]: No need for the "update" part (which is more useful when you don't have the full content when you start). IOW, you can just pass the encoded name when creating the hasher...

```
>>> hashlib.sha1(b"asd").hexdigest()
'f10e2821bbbea527ea02200352313bc059445190'
```

File: tests/unit/lxd/test_lxd_instance.py
Comment (facundobatista) [inline tests/unit/lxd/test_lxd_instance.py:915]: this is what I meant above, this is a totally valid name!

File: craft_providers/lxd/lxd_instance.py
Comment (facundobatista) [inline craft_providers/lxd/lxd_instance.py:81]: Probably this function should log (in debug) the final name, converted or not! because developers may want to "discover" the name of the instance for cases when there is need to handle them directly through LXC. What do you think?

**Review by facundobatista — CHANGES_REQUESTED** [review]
Great work!! Added some details and a couple of questions...

File: craft_providers/lxd/lxd_instance.py
Comment (mr-cal) [reply craft_providers/lxd/lxd_instance.py:81]: Yep, great idea! Added logging

File: tests/unit/lxd/test_lxd_instance.py
Comment (mr-cal) [reply tests/unit/lxd/test_lxd_instance.py]: Whoops, thanks for catching.  This was an accidental oversight.  Fixed!

File: craft_providers/lxd/lxd_instance.py
Comment (mr-cal) [reply craft_providers/lxd/lxd_instance.py]: Whoops, thanks for catching this too.  Fixed!

File: craft_providers/lxd/lxd_instance.py
Comment (facundobatista) [inline craft_providers/lxd/lxd_instance.py:114]: Let's isolate the rest of the code from this regex, just do

```
valid_name = trimmed_name.group("valid_part_of_name")
```

... and use that name after this line.

File: craft_providers/lxd/lxd_instance.py
Comment (facundobatista) [inline craft_providers/lxd/lxd_instance.py:132]: to what? you're missing something to replace the `%s`... also you probably should use `%r` here.

btw, you're not testing this log call, right?

**Review by facundobatista — CHANGES_REQUESTED** [review]
two small things!

File: craft_providers/lxd/lxd_instance.py
Comment (mr-cal) [reply craft_providers/lxd/lxd_instance.py]: Added the variable and a test.  Today I learned `assert Exact(...) in logs.debug`

File: craft_providers/lxd/lxd_instance.py
Comment (mr-cal) [reply craft_providers/lxd/lxd_instance.py:114]: +1

**Comment by mr-cal** [comment]
Changing the instance name has uncovered the next issue: `hostname`.  The `hostname` has a similar 63 max character limitation, but it's configured by the Base object, not the Executor object.

Previously, packing a charm with the name `this-is-35-characters-xxxxxxxxxxxxx` produced this error:
```
2022-08-05 15:47:58.902 * Command exit code: 1
2022-08-05 15:47:58.902 * Command output: b'Creating charmcraft-this-is-35-characters-xxxxxxxxxxxxx-21105008-0-0-amd64\n'
2022-08-05 15:47:58.902 * Command standard error output: b'Error: Invalid instance name: Name must be 1-63 characters long\n
```

Using this PR, charmcraft produces this error:
```
2022-08-05 15:48:36.933 Failed to set hostname.
* Command that failed: 'lxc --project charmcraft exec local:charmcraft-this-is-35-characters-xxxxxxx-95e3d232132477219c4f -- env PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin CHARMCRAFT_MANAGED_MODE=1 hostname -F /etc/hostname'
* Command exit code: 1
* Command standard error output: b'hostname: name too long\n'
```

Snapcraft has the same behavior.  I haven't thought of a solution that doesn't involve a breaking change, but I'll test it more.

**Review by facundobatista — APPROVED** [review]
Wonderful, thanks!

**Comment by mr-cal** [comment]
Thanks for the approval!
I'm going to hold this PR as a draft, because it is breaking the command `*craft clean` in all craft applications.
Once that issue is solved, I will land this PR.

**Pull request merged**
